scipy/optimize/optimize.py

```diff
@@ -953,6 +954,22 @@ def check_grad(func, grad, x0, *args, **kwargs):
     epsilon : float, optional
         Step size used for the finite difference approximation. It defaults to
         ``sqrt(np.finfo(float).eps)``, which is approximately 1.49e-08.
+    random_projection : bool, optional
+        If set to ``True``, then gradients along a random vector 
+        are used to check `grad` against forward difference approximation 
+        using `func`. By default it is ``False``.
```

Comment:
I know I suggested it, but I wonder if "random projection" is the best name here. We're really calculating a gradient in a random direction, so maybe "random direction" would be appropriate?

To retain more flexibility and perhaps make uses of the function more descriptive, rather than using a boolean we could make this a string option:

- `direction='all'`: checks gradient in all directions
- `direction='random'`: checks gradient only in a random direction

Reply:
>  so maybe "random direction" would be appropriate?

I agree here.



> To retain more flexibility and perhaps make uses of the function more descriptive, rather than using a boolean we could make this a string option:
> 
>     * `direction='all'`: checks gradient in all directions
> 
>     * `direction='random'`: checks gradient only in a random direction

Well, this reminded me of never copy discussions in `numpy`. Strings are better here because, we will stay in Python so input validation is easy, no boolean conversions because `direction` will never go out of `check_grad`. I wil go with `direction='all'` and `direction='random'` with default being `all` so that API doesn't break.

Reply:
I have made the change. Let me know if it looks good.

Reply:
Hi @shoyer Is this change good. I have updated the API to use, `direction='all'` and `direction='random'`. Please let me know if any changes are required. Thanks.